### PR TITLE
JIT: Clean up `RefPosition` "killed registers" storage

### PR DESCRIPTION
**Pull request by jakobbotsch**
Avoid some ifdefs and rename the field to reflect more specifically what it represents.

Also avoid calling `getWeight` entirely for `RefTypeKill` ref positions, even for dumping. Dumping this info does not make much sense since there was a requirement that it was never called in non-dump paths anyway.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
The crash this fixes happens here in builds without `HAS_MORE_THAN_64_REGISTERS`:
https://github.com/dotnet/runtime/blob/9df96f9ffe9accd08bcd1d195c8ee118ffea1652/src/coreclr/jit/lsra.cpp#L10459

`getRegisterType()` dereferences `referent` which is null for `RefTypeKill`.

**Comment by jakobbotsch** [comment]
[No asm diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=709792&view=ms.vss-build-web.run-extensions-tab). Some large Linux-hosted JIT TP regressions that I need to look into. The Windows-hosted JIT has negligible TP regressions.

**Comment by jakobbotsch** [comment]
Looks like #103573 will fix the JITDUMP crash as a separate change.

**Comment by jakobbotsch** [comment]
The linux-hosted JIT regressions are just the LTO mismatch when PGO is disabled that is being discussed in #103058:
```
Base: 50016445039, Diff: 50115771931, +0.1986%

84928137 : +53461.67% : 74.74% : +0.1698% : _ZNK9regMaskTP14IsRegNumInMaskE15_regNumber_enum               
20940077 : +8.07%     : 18.43% : +0.0419% : _ZN10LinearScan12processKillsEP11RefPosition                   
309229   : +1.64%     : 0.27%  : +0.0006% : _ZN10LinearScan24allocateRegistersMinimalEv                    
297458   : +0.33%     : 0.26%  : +0.0006% : _ZN10LinearScan25buildKillPositionsForNodeEP7GenTreej9regMaskTP
-7146243 : -0.65%     : 6.29%  : -0.0143% : _ZN10LinearScan17allocateRegistersEv                           
```

File: src/coreclr/jit/lsra.cpp [hunk 1]
```diff
@@ -10420,33 +10395,31 @@ void RefPosition::dump(LinearScan* linearScan)
...
     if (refType == RefTypeKill)
     {
-        linearScan->compiler->dumpRegMask(getKillRegisterAssignment());
+        linearScan->compiler->dumpRegMask(getKilledRegisters());
     }
     else
-#endif // HAS_MORE_THAN_64_REGISTERS
     {
         var_types type = TYP_UNKNOWN;
         if ((refType == RefTypeBB) || (refType == RefTypeKillGCRefs) || (refType == RefTypeKill))
```
Comment (jakobbotsch) [inline src/coreclr/jit/lsra.cpp:10425]:
```suggestion
        if ((refType == RefTypeBB) || (refType == RefTypeKillGCRefs))
```

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @kunalspathak 

[No asm diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=712546&view=ms.vss-build-web.run-extensions-tab). linux hosted regressions due to what I mentioned above. Some minor MinOpts TP regressions (up to 0.04% in the contexts with many MinOpts contexts). I think it's worth it to avoid the ifdefs and avoid bugs like the one fixed by #103573.

File: src/coreclr/jit/lsra.cpp [hunk 2]
```diff
@@ -178,22 +178,16 @@ void lsraAssignRegToTree(GenTree* tree, regNumber reg, unsigned regIdx)
 //
 // Returns:
 //    Weight of ref position.
-weight_t LinearScan::getWeight(RefPosition* refPos DEBUG_ARG(bool forDump))
+weight_t LinearScan::getWeight(RefPosition* refPos)
 {
+    // RefTypeKill does not have a valid treeNode field, but we do not expect
```
Comment (kunalspathak) [inline src/coreclr/jit/lsra.cpp:183]:
```suggestion
    // RefTypeKill does not have a valid treeNode field, so we do not expect
```

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks for cleaning it up. The new name `killedRegisters` is much better.

**Pull request merged**